c2rust-analyze/src/labeled_ty.rs

```diff
@@ -224,30 +224,28 @@ impl<'tcx, L: Copy> LabeledTyCtxt<'tcx, L> {
     /// Panics if `lty` contains a reference to a type parameter that is past the end of `substs`
     /// (usually this means the caller is providing the wrong list of type arguments as `substs`).
     ///
-    /// TODO: This produces a `LabeledTy` with the right structure, but doesn't actually do
-    /// substitution on the underlying `Ty`s!  This means if you substitute `u32` for `T`, you can
-    /// end up with a `LabeledTy` whose `ty` is `S<T>`, but whose args are `[u32]`.  By some
-    /// miracle, this hasn't broken anything yet, but we may need to fix it eventually.
-    #[allow(dead_code)]
-    pub fn subst(
-        &self,
-        lty: LabeledTy<'tcx, L>,
-        substs: &[LabeledTy<'tcx, L>],
-    ) -> LabeledTy<'tcx, L> {
+    /// This produces a [`LabeledTy`] with the right structure and also substitutes the
+    /// underlying [`Ty`]s
+    pub fn subst(&self, lty: LabeledTy<'tcx, L>, substs: SubstsRef<'tcx>) -> LabeledTy<'tcx, L> {
+        let substituted_ty = EarlyBinder(lty.ty).subst(self.tcx, substs);
+
         if let TyKind::Param(ref ty) = lty.ty.kind() {
-            if let Some(p) = substs.get(ty.index as usize) {
-                return p;
+            if substs.get(ty.index as usize).is_some() {
+                return self.mk(substituted_ty, &[], lty.label);
             }
         }
 
-        self.mk(lty.ty, self.subst_slice(lty.args, substs), lty.label)
+        let substituted_args = self.subst_slice(lty.args, substs);
+        debug_assert_eq!(substituted_args.len(), lty.args.len());
```

Comment:
This assert only checks that `self.subst_slice` returns a slice of the same length as its input, which is not a mistake I think we're likely to make.

The sort of bug I want to defend against is this: suppose we call rustc's `subst` and get `(i32, u8)`, and then we do substitution on our args and get the slice `[u8, i32]`.  This might happen due to some unforeseen in the way rustc's `subst` works, or due to a mismatch between the way our `subst` and rustc's `subst` handles `SubstRef`s.  Now we have an `LTy` where `lty.ty` is `(i32, u8)` but `lty.args` is `[u8, i32]` - this is not a valid `LTy`.

Maybe the right way to go about this is to write a function that checks that an `LTy` is well-formed - that is, check that `lty.args[i].ty` is equal to the i'th "argument" (as defined by `LabeledTyCtxt::label`) of the underlying `lty.ty` - and `debug_assert!` before returning from this function that the `LTy` we're returning really is well-formed.

Reply:
I think this should be covered [here](https://github.com/immunant/c2rust/commit/a7152bc6957456c8e6da4f4857e1c67ab7e62316#diff-85be9e434a05c5ae5841fcd74ecf82b89c062c10a4dbd858135f4c7781007d40R248) but let me know if I've gotten that wrong

Reply:
The latest `subst` implementation in this PR still produces wrong results.  However, when I tried to fix it myself, I discovered that the `subst` API is also broken - the `lty` and `substs` arguments it currently receives aren't sufficient to perform the right substitution.  The issue is that the index in `TyKind::Param` is an index into the full generic argument list, which includes not only types but also regions and consts, while `substs` is a list covering only the types, and we don't know how indices into the two lists match up.  For example, if `substs = [A, B]` and we see a `Param` with index 1, it might be that we're substituting into `struct Foo<T, U>` where `Param(1)` is `U` and we should return the type `B`, or it might be that we're substituting into `struct<'a, T, U>` where `Param(1)` is `T` and we should return the type `A`.

I think the thing to do here might be to revert to the old `subst` implementation (which is a bit simpler) but add an assert that `substs.len() == 0` so that the current bugs can't be triggered.  Then we can fix it in a separate PR after figuring out exactly what information it needs and how to get it.

Reply:
OBE in 2cc47277